torch/csrc/autograd/compiled_autograd.h

```diff
@@ -0,0 +1,585 @@
+#pragma once
+#include <c10/core/impl/TorchDispatchModeTLS.h>
+#include <torch/csrc/autograd/engine.h>
+#include <torch/csrc/utils/python_stub.h>
+#include <torch/csrc/utils/torch_dispatch_mode.h>
+#include <iostream>
+#include <typeindex>
+#include <vector>
+
+namespace torch {
+namespace autograd {
+namespace generated {
+struct TypeAndSize;
+}
+
+struct SizeInput {
+  enum DynType : uint8_t { STATIC = 0, DYNAMIC = 1 };
+  DynType dyn_type;
+  int64_t value;
+};
+
+struct CacheKeyBuffer {
+  CacheKeyBuffer() : data(nullptr) {}
+  CacheKeyBuffer(const char* key, uint16_t len) : data(new char[len]) {
+    memcpy(data, key, len);
+  }
+  CacheKeyBuffer(CacheKeyBuffer&& other)
+      : data(std::exchange(other.data, nullptr)) {}
+  ~CacheKeyBuffer() {
+    if (data != nullptr) {
+      delete[] data;
+    }
+  }
+  CacheKeyBuffer(const CacheKeyBuffer& other) = delete;
+  CacheKeyBuffer& operator=(const CacheKeyBuffer&) = delete;
+
+  char* data;
+};
+
+struct CacheKey {
+  CacheKey(const std::type_index& ntype, const char* key, uint16_t len)
+      : node_type(ntype), key_size(len), key(key) {}
+
+  bool operator<(const CacheKey& other) const {
+    if (node_type != other.node_type)
+      return node_type < other.node_type;
+    if (key_size != other.key_size)
+      return key_size < other.key_size;
+    return memcmp(key, other.key, key_size) < 0;
+  }
+
+  bool operator==(const CacheKey& other) const {
+    return node_type == other.node_type && key_size == other.key_size &&
+        memcmp(key, other.key, key_size) == 0;
+  }
+
+  size_t hash() const {
+    // don't bother hashing the key data, common case 1 cache entry per node
+    return std::hash<std::type_index>()(node_type) ^ key_size;
+  }
+
+  std::type_index node_type;
+  uint16_t key_size;
+  const char* key;
+};
+
+struct OutputRef {
+  OutputRef() : node_id(-1), index(-1) {}
+  OutputRef(int node_, int index_) : node_id(node_), index(index_) {}
+
+  bool is_set() const {
+    return this->node_id >= 0;
+  }
+
+  int node_id;
+  int index;
+};
+
+struct NodeCall {
+  NodeCall(const std::shared_ptr<Node>& node_)
+      : node(node_), input_refs(node_->num_inputs()) {}
+
+  NodeCall(const std::shared_ptr<Node>& node_, int num_inputs_)
+      : node(node_), input_refs(num_inputs_) {}
+
+  OutputRef& operator[](size_t pos) {
+    return input_refs[pos];
+  }
+
+  size_t size() const {
+    return input_refs.size();
+  }
+
+  void mark_output(int input_nr, int output_idx) {
+    graph_output.emplace_back(std::make_pair(input_nr, output_idx));
+  }
+
+  std::shared_ptr<Node> node;
```

Comment:
shared_ptr lifecycles are ambigious and a little code smell - what is the ownership story for the node?

Reply:
It needs to be a `shared_ptr` because this is getting passed to SavedVariable::unpack that require a `std::shared_ptr<Node>`.

Reply:
Ah lame, but it is what it is :)